### ISSUE TITLE
Vocabulary boosting / CTC keyword spotting on the unified Parakeet path?

### ISSUE THREAD

**Issue by darylthornhill** [FluidInference/FluidAudio#851]
Hi - first, thanks for FluidAudio. We use it for fully on-device meeting
transcription in a macOS app, currently on 0.15.5.

**What we do today**

We run `StreamingAsrManager` with Parakeet TDT v3 and feed it a per-meeting
custom vocabulary (product names, clinical terms, people's names) via:

```swift
try await manager.configureVocabularyBoosting(
    vocabulary: context,
    ctcModels: ctcModels   // CtcModels.downloadAndLoad()
)
```

That boosting matters a lot for us - without it, domain terms come back
mangled and everything downstream inherits the error.

**What we would like to do**

Move to `parakeet-unified-en-0.6b`, mainly for the reported AMI and
SPGISpeech numbers, since ours is exactly the meetings-plus-jargon case.

**Where we got stuck**

Reading the 0.15.5 source, `configureVocabularyBoosting` appears to be defined
only on `SlidingWindowAsrManager`:

- `Sources/FluidAudio/ASR/Parakeet/SlidingWindow/SlidingWindowAsrManager.swift`

and we could not find any equivalent on `UnifiedAsrManager` or
`StreamingUnifiedAsrManager` under `Sources/FluidAudio/ASR/Parakeet/Unified/`.

At the same time, `CtcKeywordSpotter` carries two methods whose docs
explicitly mention the unified path:

- "Use this when CTC logProbs are already available (e.g. from a unified
  Preprocessor ...)"
- "Use this to post-process raw logits from a unified Preprocessor before
  passing to ..."

So the primitives look like they anticipate this, but we could not find the
wiring between the unified managers and the keyword spotter.

**Questions**

1. Is vocabulary boosting on the unified path supported today in a way we have
   missed? If so, is there an example we could follow?
2. If not, is it on the roadmap? We are happy to wait rather than build
   something out-of-tree that then diverges.
3. If it is not planned soon: is exposing the unified preprocessor's CTC
   logits and driving `CtcKeywordSpotter` ourselves a reasonable approach, or
   is there a reason that will not align correctly with the unified encoder?
   We noticed issue #435 was closed without a fix and want to avoid repeating
   whatever went wrong there.

Happy to test any branch on real meeting audio and report back - we have a
fixture-based eval harness set up, so we can give you before/after numbers on
our own recordings.

Thanks!

**Comment by codedbybencorp** [comment]
Following up on this with measurements from 0.15.6, in case they are useful —
and in case we are holding it wrong, which is the more likely reading of a
result this clean.

Terms and audio are from a private vocabulary, so they are described rather
than quoted below; I would rather withhold them than paste substitutes the
model never actually emitted. Happy to share the clip and the exact list
privately with a maintainer.

## Setup

macOS 26.6.2, M3 Pro, FluidAudio 0.15.6 (exact pin), models resolved from
`~/Library/Application Support/FluidAudio/Models`. One 7.1 s clip, single
speaker, clean audio, naming five domain nouns — two place names, one
lowercase internal tool name, one personal name, one well-known BI product —
inside an ordinary sentence. Vocabulary context of 15 terms, including all
five.

## Result

**0 of 5 terms recovered, in every configuration**, and more usefully:

| path | boosted vs unboosted |
|---|---|
| `StreamingUnifiedAsrManager` | **byte-identical** |
| `SlidingWindowAsrManager` (parakeet-tdt-0.6b-v3) | **byte-identical** |

Not "boosting helped a little" — `configureVocabularyBoosting` makes no
difference to the output string at all, on either manager. The two managers
differ from each other (the sliding-window one gets one of the five place
names right on its own, unboosted), so the models are certainly doing
different work; the boosting call is what appears to be a no-op.

The three error shapes, since they may matter for whether the spotter ever
proposes a span:

1. a syllable transposition within a name (edit distance 2 on a 10-character
   term, so string similarity should be high)
2. a single lowercase compound tool name split into two words with different
   casing
3. a two-word product name that survives correctly *without* boosting

## What we tried

- Both managers, `encoderPrecision: .int8`, 1 s `AVAudioPCMBuffer` chunks at
  16 kHz mono Float32.
- `CtcModels.downloadAndLoad()` succeeds, `parakeet-ctc-110m-coreml` is on
  disk, and `configureVocabularyBoosting` returns without throwing.
- Default thresholds, then deliberately permissive ones:
  `alpha: 2.0, minCtcScore: -20.0, minSimilarity: 0.30,
  minCombinedConfidence: 0.30, minTermLength: 3`. Identical output at both.
- Sliding window: a fresh `SlidingWindowAsrManager` per run, and
  `transcriptionUpdates` subscribed before `startStreaming`, given that its
  input stream is single-use.
- Unified: `reset()` before each run, with `configureVocabularyBoosting`
  called after `reset()` so the retained-audio clock anchors at position 0.

## Questions

1. Is a 7.1 s clip simply too short? `vocabSegmentSeconds` is 15.0, and while
   `finish()` calls `maybeRescoreSegment(force: true)`, we may be misreading
   that path. Is there a minimum audio length below which a rescore cannot
   fire in practice?
2. Should terms carry `aliases` (spoken forms) rather than surface forms
   alone, for names the model has no lexicon entry for?
3. Is `tokenIds` / `ctcTokenIds` pre-tokenization effectively required rather
   than optional for out-of-lexicon terms?

Happy to test any branch and report back on the same clip.

For context on the stakes: Apple's `DictationTranscriber` with
`AnalysisContext.contextualStrings` recovers all five terms correctly on this
audio, while scoring far worse on ordinary prose — mean WER 0.157 against
parakeet's 0.000 on two other recordings of the same speaker, scored against a
third-party transcript of the same audio. We are currently shipping both
engines and asking the user to choose, which we would rather not do.

**Comment by Alex-Wengg** [comment]
Thanks, this was a real report and you weren't holding it wrong. Reproduced on 0.15.6 with a public earnings22 clip. Three separate causes:

1. `CustomVocabularyContext(terms: [CustomVocabularyTerm(text: ...)])` never tokenizes, despite what the doc says. The spotter and rescorer silently skip terms without `ctcTokenIds`, so boosting is a no-op on every engine. Only `loadWithCtcTokens(from:)` tokenizes today. As a workaround, tokenize each term with `CtcTokenizer.load()` and `encode(_:)` and pass the result as `ctcTokenIds`. That's your Q3, answered yes, and it's a bug in our API and docs.
2. `SlidingWindowAsrManager` only rescores confirmed windows, which need 10 s of context by default. A 7.1 s clip never confirms, so it's never rescored even with tokenized terms. Batch mode has no such gate.
3. With boosting on, an unconfirmed trailing window can overwrite the previous window's rescored text in the sliding-window manager. On a 15 s clip that returned an empty transcript.

The unified streaming engine with tokenized terms corrected a 9 s clip here, so once the terms are tokenized that path should work on your audio. Fixes for all three are coming. If you can share the clip I'll run it against the branch.

To share the clip privately, DM me on the FluidInference Discord (link in the README) or email hanweng9@gmail.com.

**Comment by Alex-Wengg** [comment]
Closing as fixed by #898 (merged as 07c8903b). Three defects were behind the byte-identical result:

1. Terms built in code carried no CTC token IDs and every consumer skipped them silently. `configureVocabularyBoosting` / `VocabularyBoostingSession` now tokenize such terms with the CTC tokenizer, so the documented `CustomVocabularyContext(terms: [CustomVocabularyTerm(text:)])` form works as written on every engine. Pre-tokenized terms are used as-is; the workaround from my earlier comment is no longer needed.
2. `SlidingWindowAsrManager` only rescored windows that confirmed at decode time (needs 10 s of context by default), so a 7.1 s clip was never rescored. Every window is rescored now.
3. With boosting on, an unconfirmed trailing window overwrote the previous window's text and `finish()` could return an empty transcript. Fixed.

Verified on real audio: a 9 s clip now corrects the term in streaming, a 15 s clip that returned empty is complete and corrected, batch and unboosted output unchanged. An end-to-end test feeds untokenized in-code terms through streaming on a public recording with no window allowed to confirm, and runs in CI.

@codedbybencorp a re-run of your 7.1 s clip on main would be the confirming data point; the private-clip offer via Discord or email stands. One thing to be aware of on the sliding-window path: its default rescue floors can over-fire on short unrelated words (tracked in #899); `StreamingUnifiedAsrManager` already uses the stricter floors.
